html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,134 @@
+
+//Has the user entered data with length > 0 & <= 256?
+if (($user_data) && (strlen($user_data) <= 256)) {
+	// Retrieve the logged in user's UserID.
+	$userid = $user->id;
+
+	// Create the message which will be signed.
+	$message_data = "$userid $user_data";
+
+	// Create a digest of the message_data.
+	// Without the digest, the openssl_sign function fails to sign lower case characters.
+	// We can prove the integrity of the message with a digest.
+	$message_digest = openssl_digest ($message_data , "sha512"); 
+
+	// How to generate required keys in /project/keys/ folder:
+	// openssl genpkey -algorithm RSA -out external_system_key_private.pem -pkeyopt rsa_keygen_bits:2048
+	// openssl rsa -pubout -in external_system_key_private.pem -out external_system_key_public.pem
+
+	$private_key_path = "file:///root/project/keys/external_system_key_private.pem";
+	$public_key_path = "file:///root/project/keys/external_system_key_public.pem";
+
+	// Check that the private key file exists where specified. If not, redirect to error page.
+	if (!file_exists($private_key_path)) {
+		error_page(tra("The private key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Check that the public key file exists where specified. If not, redirect to error page.
+	if (!file_exists($public_key_path)) {
+		error_page(tra("The public key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Fetch private key from file
+	$private_key_pem = openssl_pkey_get_private($private_key_path);
+
+	// Compute signature using the SHA512 signing algorithm
+	openssl_sign($message_digest, $signature, $private_key_pem, OPENSSL_ALGO_SHA512);
+
+	// Free the private key from memory
+	openssl_free_key($private_key_pem);
+	
+	// Check that the signature exists
+	if ($signature) {
+		// The signature was successfully created.
+
+		// fetch public key into memory
+		$pubkeyid = openssl_pkey_get_public($public_key_path);
+
+		// Base64 encode the generated signature to enable safe output to text file.
+		$base64_sig = base64_encode($signature);
+
+		// Decode base64 sig for use in sig_verification
+		$decoded_sig = base64_decode($base64_sig);
+
+		// Verify that the generated signature against the original data, using the public key.
+		$sig_verification = openssl_verify($message_digest, $decoded_sig, $pubkeyid, OPENSSL_ALGO_SHA512);
+		
+		// Free the public key from memory
+		openssl_free_key($pubkeyid);
+
+		// Generate an auto-download signature text file prompt
+		  // TODO: Verify this txt file auto-download works on many browsers
+		  // Question: Does outputting the signature as plain text corrupt the signature? Verify manually!
+		  // Verified: Firefox
+		header("Content-Disposition: attachment; filename=signature.txt");
+		header("Content-type: text/plain");
+
+		// Start outputting text to the auto-download txt file
+		echo "-----BEGIN USERID VERIFICATION-----";
+
+			echo "\nUSER_ID: $userid";
+			echo "\nUSER_DATA: $user_data";
```

Comment:
Why print the user data when that is contained in message?

Reply:
Left behind debugging info for the lower-case (non-digested) user_data signing, both the USER_ID and USER_DATA could be removed since the MSG line stores both with a space delimiter. Sound good?